File: src/ui/tray.rs
//! Tray icon registration, theme-aware glyph, and balloon notifications.
//!
//! The glyph is rendered at runtime by [`crate::logo`] at `16 * dpi / 96` px:
//! white on the default dark taskbar, dark on a light one.

use std::sync::atomic::{AtomicIsize, Ordering};

use windows::Win32::{
    Foundation::HWND,
    Graphics::Gdi::{CreateBitmap, DeleteObject},
    UI::{
        HiDpi::GetDpiForWindow,
        Shell::{
            Shell_NotifyIconW, NIF_ICON, NIF_INFO, NIF_MESSAGE, NIF_TIP, NIIF_INFO, NIIF_WARNING,
            NIM_ADD, NIM_DELETE, NIM_MODIFY, NIM_SETVERSION, NOTIFYICONDATAW, NOTIFYICONDATAW_0,
        },
        WindowsAndMessaging::{CreateIconIndirect, DestroyIcon, HICON, ICONINFO},
    },
};

use super::imp::WM_TRAY_CALLBACK;
use crate::logo;

const TRAY_ID: u32 = 1;
/// `NOTIFYICON_VERSION` (3): keeps the classic `WM_LBUTTONUP`/`WM_RBUTTONUP`
/// callback semantics AND makes the shell deliver balloon events
/// (`NIN_BALLOONUSERCLICK`). Without `NIM_SETVERSION` the shell stays on
/// version 0, which never sends balloon clicks — the update balloon would
/// look clickable but do nothing.
const NOTIFYICON_VERSION_3: u32 = 3;
/// The current glyph handle, so a theme flip can destroy the old one.
static CURRENT_ICON: AtomicIsize = AtomicIsize::new(0);

/// Registers the tray icon.
///
/// # Errors
/// Returns a message when the glyph or the shell registration fails.
pub(super) fn add(hwnd: HWND, light_taskbar: bool) -> Result<(), String> {
    let icon = glyph(hwnd, light_taskbar).ok_or("render tray glyph")?;
    let mut data = base_data(hwnd);
    data.uFlags = NIF_ICON | NIF_MESSAGE | NIF_TIP;
    data.uCallbackMessage = WM_TRAY_CALLBACK;
    data.hIcon = icon;
    copy_wide(&mut data.szTip, "Tokitoki");
    // SAFETY: fully initialized NOTIFYICONDATAW for our own window.
    if !unsafe { Shell_NotifyIconW(NIM_ADD, &raw const data) }.as_bool() {
        return Err("Shell_NotifyIconW(NIM_ADD) failed".to_owned());
    }
    let mut version_data = base_data(hwnd);
    version_data.Anonymous = NOTIFYICONDATAW_0 {
        uVersion: NOTIFYICON_VERSION_3,
    };
    // SAFETY: opting the freshly added icon into version-3 callbacks.
    unsafe {
        let _ = Shell_NotifyIconW(NIM_SETVERSION, &raw const version_data);
    }
    swap_current(icon);
    Ok(())
}

/// Swaps in a re-rendered glyph after a taskbar theme change.
pub(super) fn refresh_icon(hwnd: HWND, light_taskbar: bool) {
    let Some(icon) = glyph(hwnd, light_taskbar) else {
        return;
    };
    let mut data = base_data(hwnd);
    data.uFlags = NIF_ICON;
    data.hIcon = icon;
    // SAFETY: modifying our registered tray icon.
    if unsafe { Shell_NotifyIconW(NIM_MODIFY, &raw const data) }.as_bool() {
        swap_current(icon);
    } else {
        destroy(icon);
    }
}

/// Shows a balloon notification.
pub(super) fn balloon(hwnd: HWND, title: &str, text: &str, warning: bool) {
    let mut data = base_data(hwnd);
    data.uFlags = NIF_INFO;
    data.dwInfoFlags = if warning { NIIF_WARNING } else { NIIF_INFO };
    copy_wide(&mut data.szInfoTitle, title);
    copy_wide(&mut data.szInfo, text);
    // SAFETY: modifying our registered tray icon.
    unsafe {
        let _ = Shell_NotifyIconW(NIM_MODIFY, &raw const data);
    }
}

/// Removes the tray icon and frees the glyph.
pub(super) fn remove(hwnd: HWND) {
    let data = base_data(hwnd);
    // SAFETY: deleting our registered tray icon.
    unsafe {
        let _ = Shell_NotifyIconW(NIM_DELETE, &raw const data);
    }
    swap_current(HICON(std::ptr::null_mut()));
}

fn base_data(hwnd: HWND) -> NOTIFYICONDATAW {
    NOTIFYICONDATAW {
        cbSize: u32::try_from(std::mem::size_of::<NOTIFYICONDATAW>()).unwrap_or(0),
        hWnd: hwnd,
        uID: TRAY_ID,
        ..Default::default()
    }
}

fn glyph(hwnd: HWND, light_taskbar: bool) -> Option<HICON> {
    // SAFETY: querying DPI of our own window.
    let dpi = unsafe { GetDpiForWindow(hwnd) }.max(96);
    let size = 16 * dpi / 96;
    let color = if light_taskbar {
        logo::DARK
    } else {
        logo::WHITE
    };
    hicon(&logo::mark(size, color))
}

/// Builds an `HICON` from a straight-alpha RGBA image via a 32bpp BGRA
/// bitmap plus an empty monochrome mask.
fn hicon(image: &logo::Image) -> Option<HICON> {
    let bgra: Vec<u8> = image
        .rgba
        .chunks_exact(4)
        .flat_map(|px| [px[2], px[1], px[0], px[3]])
        .collect();
    let side = i32::try_from(image.size).ok()?;
    // SAFETY: the pixel buffer outlives both CreateBitmap calls; GDI objects
    // are released on every path.
    unsafe {
        let color = CreateBitmap(side, side, 1, 32, Some(bgra.as_ptr().cast()));
        let mask = CreateBitmap(side, side, 1, 1, None);
        let info = ICONINFO {
            fIcon: true.into(),
            xHotspot: 0,
            yHotspot: 0,
            hbmMask: mask,
            hbmColor: color,
        };
        let icon = CreateIconIndirect(&raw const info).ok();
        let _ = DeleteObject(color.into());
        let _ = DeleteObject(mask.into());
        icon
    }
}

fn swap_current(new: HICON) {
    let old = CURRENT_ICON.swap(new.0 as isize, Ordering::AcqRel);
    destroy(HICON(old as *mut core::ffi::c_void));
}

fn destroy(icon: HICON) {
    if !icon.0.is_null() {
        // SAFETY: destroying an icon we created and no longer use.
        unsafe {
            let _ = DestroyIcon(icon);
        }
    }
}

fn copy_wide(dest: &mut [u16], text: &str) {
    let encoded: Vec<u16> = text.encode_utf16().take(dest.len() - 1).collect();
    dest[..encoded.len()].copy_from_slice(&encoded);
    dest[encoded.len()] = 0;
}

File: src/ui/tray.c
/* Tray icon registration, theme-aware glyph, balloon notifications.
 * The glyph is rendered at runtime at 16 * dpi / 96 px: white on the default
 * dark taskbar, dark on a light one.
 *
 * NIM_SETVERSION(NOTIFYICON_VERSION) after every NIM_ADD is load-bearing:
 * without it the shell uses version-0 semantics and never delivers
 * NIN_BALLOONUSERCLICK, so update balloons would be unclickable. */
#include "ui/ui.h"

#include "logo.h"

#include <shellapi.h>

#include <stdlib.h>
#include <string.h>

#pragma comment(lib, "shell32.lib")

#define TRAY_ID 1

static HICON g_current_icon;

static HICON glyph_icon(HWND hwnd, bool light_taskbar) {
    UINT dpi = GetDpiForWindow(hwnd);
    if (dpi < 96) {
        dpi = 96;
    }
    UINT size = 16 * dpi / 96;
    uint8_t *rgba = malloc((size_t)size * size * 4);
    if (!rgba) {
        return NULL;
    }
    logo_mark(size, !light_taskbar, rgba);

    /* RGBA (straight alpha) -> BGRA for the 32bpp color bitmap. */
    for (size_t i = 0; i < (size_t)size * size * 4; i += 4) {
        uint8_t r = rgba[i];
        rgba[i] = rgba[i + 2];
        rgba[i + 2] = r;
    }
    HBITMAP color = CreateBitmap((int)size, (int)size, 1, 32, rgba);
    HBITMAP mask = CreateBitmap((int)size, (int)size, 1, 1, NULL);
    free(rgba);
    if (!color || !mask) {
        if (color) {
            DeleteObject(color);
        }
        if (mask) {
            DeleteObject(mask);
        }
        return NULL;
    }
    ICONINFO info = {TRUE, 0, 0, mask, color};
    HICON icon = CreateIconIndirect(&info);
    DeleteObject(color);
    DeleteObject(mask);
    return icon;
}

static void base_data(HWND hwnd, NOTIFYICONDATAW *data) {
    memset(data, 0, sizeof(*data));
    data->cbSize = sizeof(*data);
    data->hWnd = hwnd;
    data->uID = TRAY_ID;
}

static void swap_current(HICON icon) {
    if (g_current_icon) {
        DestroyIcon(g_current_icon);
    }
    g_current_icon = icon;
}

bool tray_add(HWND hwnd, bool light_taskbar) {
    HICON icon = glyph_icon(hwnd, light_taskbar);
    if (!icon) {
        return false;
    }
    NOTIFYICONDATAW data;
    base_data(hwnd, &data);
    data.uFlags = NIF_ICON | NIF_MESSAGE | NIF_TIP;
    data.uCallbackMessage = WM_TRAY_CALLBACK;
    data.hIcon = icon;
    wcscpy_s(data.szTip, ARRAYSIZE(data.szTip), L"Tokitoki");
    if (!Shell_NotifyIconW(NIM_ADD, &data)) {
        DestroyIcon(icon);
        return false;
    }
    swap_current(icon);

    data.uFlags = 0;
    data.uVersion = NOTIFYICON_VERSION;
    Shell_NotifyIconW(NIM_SETVERSION, &data);
    return true;
}

void tray_refresh_icon(HWND hwnd, bool light_taskbar) {
    HICON icon = glyph_icon(hwnd, light_taskbar);
    if (!icon) {
        return;
    }
    NOTIFYICONDATAW data;
    base_data(hwnd, &data);
    data.uFlags = NIF_ICON;
    data.hIcon = icon;
    if (Shell_NotifyIconW(NIM_MODIFY, &data)) {
        swap_current(icon);
    } else {
        DestroyIcon(icon);
    }
}

void tray_balloon(HWND hwnd, const wchar_t *title, const wchar_t *text, bool warning) {
    NOTIFYICONDATAW data;
    base_data(hwnd, &data);
    data.uFlags = NIF_INFO;
    data.dwInfoFlags = warning ? NIIF_WARNING : NIIF_INFO;
    /* Truncate, never abort: caller text (localized CLI stderr) can exceed
     * these fixed fields, and wcscpy_s would trip the CRT invalid-parameter
     * handler and terminate the process. */
    wcsncpy_s(data.szInfoTitle, ARRAYSIZE(data.szInfoTitle), title, _TRUNCATE);
    wcsncpy_s(data.szInfo, ARRAYSIZE(data.szInfo), text, _TRUNCATE);
    Shell_NotifyIconW(NIM_MODIFY, &data);
}

void tray_remove(HWND hwnd) {
    NOTIFYICONDATAW data;
    base_data(hwnd, &data);
    Shell_NotifyIconW(NIM_DELETE, &data);
    swap_current(NULL);
}
